src/Features/CSharp/Portable/Completion/CompletionProviders/OperatorIndexerCompletionProvider.cs

```diff
@@ -0,0 +1,220 @@
+                _ => null,
+            };
+        }
+
+        internal override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, TextSpan completionListSpan, char? commitKey, bool disallowAddingImports, CancellationToken cancellationToken)
+        {
+            if (item.Properties.TryGetValue(CompletionHandlerPropertyName, out var value))
+            {
+                var completionChange = value switch
+                {
+                    CompletionHandlerConversion => await HandleConversionChangeAsync(document, item, cancellationToken).ConfigureAwait(false),
+                    CompletionHandlerIndexer => await HandleIndexerChangeAsync(document, item, cancellationToken).ConfigureAwait(false),
+                    _ => null,
+                };
+                if (completionChange is { })
+                {
+                    return completionChange;
+                }
+            }
+
+            return await base.GetChangeAsync(document, item, completionListSpan, commitKey, disallowAddingImports, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<CompletionChange?> HandleConversionChangeAsync(Document document, CompletionItem item, CancellationToken cancellationToken)
+        {
+            var position = SymbolCompletionItem.GetContextPosition(item);
+            if (!item.Properties.TryGetValue(MinimalTypeNamePropertyName, out var typeName))
+            {
+                return null;
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindTokenOnLeftOfPosition(position);
+            // syntax tree manipulations are to complicated if a mixture of conditionals is involved. Some text manipulation is easier here.
+            //                      ↓               | cursor position
+            // white?.Black.White.Black?.White      | current user input
+            // white?.Black.White.Black?.White      | rootExpression (text manipulation starts with this)
+            //       .Black.White                   | parentExpression (needed to calculate the position to insert the closing brace)
+            //                    Black             | identifier at cursor position (gets removed, because the user typed the name of a type)
+            // |----------------------|             | part to replace (TextChange.Span), if identifier is not present: ends at rootExpression.End (after White.)
+            //                   ↑                  | insert closing brace after parentExpression.Span.End
+            // ((Black)white?.Black.White).?.White  | The result. Because we removed the identifier, the remainder after the identifier may be syntactically wrong 
+            //                             ↑        | cursor after the manipulation is placed after the dot
```

Comment:
i'm not really understanding the issue.  given   `before ?. after` if we want to cast, woudn't we just convert that to `((type)before ?. after)`?

Reply:
The problem is, that given the expression:
`a?.B.$$`
we need the `a?.B` (without the second dot) as `ExpressionSyntax` to build the cast syntax, but the syntax tree contains:
* ConditionalAccessExpression (`a?.B.`)
  * Expression: IdentifierName (`a`)
  * WhenNotNull: SimpleMemberAccessExpression (`.B.`)
    * Expression: MemberBindingExpression (`.B`)

There is no matching expression. The best match is ConditionalAccessExpression `a?.B.`, but the dot after the B is hampering its use (the troublesome dot can be somewhere deeply nested in the `WhenNotNull` tree).
I came to the conclusion, that we need to manipulate the ConditionalAccessExpression to get rid of the dot, but I could'nd find a way to do it without resorting to some kind of `ToString` -> string manipulation -> reparsing. Given that, it was easier to do some simple string manipulation all the way, especially because `TextChange` requires it anyway at the end.

Is there a better way to solve this?

Reply:
yup.  string manipulation seems reasonable here.  will have to look at it to make sure it handles things properly.  one thing you might want to test is when there is explicit trivia preceding the expression (like comments). to make sure that the indices are correct.

Reply:
This is the trivia test. Is this sufficient?
https://github.com/dotnet/roslyn/blob/3009adc8351ad6dc4cfe99d8fef1906ddee2ff2c/src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ConversionCompletionProviderTests.cs#L722-L760